lib/opte/src/engine/packet.rs

```diff
@@ -3740,4 +3884,88 @@ mod test {
         // And make sure they don't include the padding bytes
         assert_eq!(ip6_hdr.pay_len(), udp_hdr.hdr_len() + body.len());
     }
+
+    fn create_linked_mblks(n: usize) -> Vec<*mut mblk_t> {
+        let mut els = vec![];
+        for i in 0..n {
+            els.push(allocb(8));
+        }
+
+        // connect the elements in a chain
+        for (lhs, rhs) in els.iter().zip(els[1..].iter()) {
+            unsafe {
+                (**lhs).b_next = *rhs;
+                (**rhs).b_prev = *lhs;
+            }
+        }
+
+        els
+    }
+
+    #[test]
+    fn chain_has_correct_ends() {
+        let mut els = create_linked_mblks(3);
+
+        let chain = unsafe { PacketChain::new(els[0]) }.unwrap();
+        let chain_inner = chain.inner.as_ref().unwrap();
+        assert_eq!(chain_inner.head.as_ptr(), els[0]);
+        assert_eq!(chain_inner.tail.as_ptr(), els[2]);
+    }
+
+    #[test]
+    fn chain_breaks_links() {
+        let els = create_linked_mblks(3);
+
+        let mut chain = unsafe { PacketChain::new(els[0]) }.unwrap();
+
+        let p0 = chain.next().unwrap();
+        assert_eq!(p0.mblk_addr(), els[0] as uintptr_t);
+        unsafe {
+            assert!((*els[0]).b_prev.is_null());
+            assert!((*els[0]).b_next.is_null());
+        }
+
+        // Chain head/tail ptrs are correct
+        let chain_inner = chain.inner.as_ref().unwrap();
+        assert_eq!(chain_inner.head.as_ptr(), els[1]);
+        assert_eq!(chain_inner.tail.as_ptr(), els[2]);
```

Comment:
Nit: we could also check that the head + tail aren't connected to anything (the tail won't have changed, but the head will have changed in the call to `next()`)
```suggestion
        assert_eq!(chain_inner.head.as_ptr(), els[1]);
        assert!((*els[1]).b_prev.is_null());
        assert_eq!(chain_inner.tail.as_ptr(), els[2]);
        assert!((*els[2]).b_next.is_null());
```

Reply:
Done.